contracts/info-type-script/src/entry/verify.rs

```diff
@@ -0,0 +1,392 @@
+use core::result::Result;
+
+use num_bigint::{BigInt, BigUint};
+use num_traits::identities::Zero;
+
+// Import CKB syscalls and structures
+// https://nervosnetwork.github.io/ckb-std/riscv64imac-unknown-none-elf/doc/ckb_std/index.html
+use share::cell::{InfoCellData, LiquidityOrderLockArgs, SUDTAmountData};
+use share::ckb_std::{
+    ckb_constants::Source,
+    ckb_types::{packed::CellOutput, prelude::*},
+    // debug,
+    high_level::{load_cell, load_cell_data, load_cell_lock_hash, QueryIter},
+};
+use share::{decode_u128, get_cell_type_hash};
+
+use crate::error::Error;
+
+const INFO_VERSION: u8 = 1;
+
+const THOUSAND: u128 = 1_000;
+const TEN_THOUSAND: u128 = 10_000;
+const SUDT_CAPACITY: u64 = 15_400_000_000;
+const INFO_CAPACITY: u64 = 21_400_000_000;
+
+pub fn liquidity_tx_verification() -> Result<(), Error> {
+    let info_in_data = InfoCellData::from_raw(&load_cell_data(0, Source::Input)?)?;
+    let pool_in_cell = load_cell(1, Source::Input)?;
+    let info_out_cell = load_cell(0, Source::Output)?;
+    let info_out_data = InfoCellData::from_raw(&load_cell_data(0, Source::Output)?)?;
+    let pool_out_cell = load_cell(1, Source::Output)?;
+    let pool_out_data = SUDTAmountData::from_raw(&load_cell_data(1, Source::Output)?)?;
+
+    let mut liquidity_sudt_type_hash = [0u8; 20];
+    liquidity_sudt_type_hash.copy_from_slice(&info_in_data.liquidity_sudt_type_hash);
+    let pool_type_hash = get_cell_type_hash!(1, Source::Input);
+
+    let ckb_reserve = info_in_data.ckb_reserve;
+    let sudt_reserve = info_in_data.sudt_reserve;
+    let total_liquidity = info_in_data.total_liquidity;
+
+    let mut pool_ckb_paid = 0;
+    let mut pool_sudt_paid = 0;
+    let mut ckb_collect = 0;
+    let mut sudt_collect = 0;
+    let mut user_liquidity_mint = 0;
+    let mut user_liquidity_burned = 0;
+    let mut base_index = 0;
+
+    for (idx, (liquidity_order_cell, raw_data)) in QueryIter::new(load_cell, Source::Input)
+        .zip(QueryIter::new(load_cell_data, Source::Input))
+        .enumerate()
+        .skip(3)
+    {
+        let liquidity_order_lock_args =
+            LiquidityOrderLockArgs::from_raw(&liquidity_order_cell.lock().args().as_slice())?;
+        if liquidity_order_lock_args.version != INFO_VERSION {
+            return Err(Error::VersionDiff);
+        }
+
+        // Todo: fix me
+        let liquidity_order_data = SUDTAmountData::from_raw(&raw_data)?;
+        let liquidity_type_hash = get_cell_type_hash!(idx, Source::Input);
+        if liquidity_order_lock_args.info_type_hash.as_ref()
+            != &get_cell_type_hash!(0, Source::Input)[0..20]
+        {
+            return Err(Error::InvalidLiquidityCell);
+        }
+
+        if info_in_data.total_liquidity == 0 {
+            if QueryIter::new(load_cell, Source::Input).count() == 4 {
+                return verify_initial_mint(
```

Comment:
Should break out loop instead of return. If we return here, then we skip info and pool verification.